### PR TITLE
Deprecates `actfw` package in favor of `actfw-core` and `actfw-raspberrypi`, keeping forward compatibility

### PR DESCRIPTION
**Pull request by laysakura**
Will separate this `actfw` package into `actfw-core` and `actfw-raspberrypi`.

- [`actfw-core`](https://pypi.org/project/actfw-core/) : Core components of actfw, which requires Linux and V4L2.
- [`actfw-raspberrypi`](https://pypi.org/project/actfw-raspberrypi/) : Raspberry Pi-specific components, which depends on [picamera](https://pypi.org/project/picamera/) and Video Core IV.

`actfw` repository remains only for forward compatibility of Actcast applications and will stop further update since v1.4.0.

## Migration plan

1. Upload `actfw v1.4.0` (intended to be the last version), `actfw-core v1.4.0`, `actfw-raspberrypi v1.4.0` to PyPI.
2. (Existing Actcast apps work with both `actfw < v1.4.0` & `actfw = v1.4.0`)
3. Guide Actcast app vendors to use `actfw-core` and `actfw-raspberrypi` in somewhere like https://actcast.io/docs/ForVendor/ApplicationDevelopment/actfw/
4. Rewrites and updates Idein's Actcast apps to use `actfw-core` and `actfw-raspberrypi`.

## Testing

On a Raspberry Pi device:

```bash
sudo apt-get update
sudo apt-get install -y python3-pip python3-pil 

pip3 install pipenv

pipenv install --pre --dev -e .  # `--pre` is not required after `actfw-core` and `actfw-raspberrypi` have non-pre release version.
pipenv run install-raspberrypi

pipenv run python <app using actfw>
```

You can use https://github.com/Idein/actcast-app-python/blob/master/example/hello for testing, for example.

## TODO

- [x] @laysakura Upload release version of `actfw-core` and `actfw-raspberrypi` to PyPI after code review.
- [x] @laysakura rerun CI (currently CI fails because `actfw-{core,raspberrypi}` are pre-release version and cannot install without `--pre` option.
- [x] @laysakura Upload `actfw` `v1.4.0a1` to TestPyPI after code review.
- [x] @laysakura Test forward compatibility with some Actcast applications.
  - using actcast-app-idein-imagenet-classification (private)

### CODE REVIEW

File: actfw/application/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from actfw_core.application import *
```
Comment (notogawa) [inline actfw/application/__init__.py:1]:
```suggestion
from actfw_core.application import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/capture/__init__.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+from actfw_core.capture import *
+from actfw_raspberrypi.capture import *
```
Comment (notogawa) [inline actfw/capture/__init__.py:2]:
```suggestion
from actfw_core.capture import *
from actfw_raspberrypi.capture import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core and actfw-raspberrypi instead"))
```

File: actfw/command_server/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from actfw_core.command_server import *
```
Comment (notogawa) [inline actfw/command_server/__init__.py:1]:
```suggestion
from actfw_core.command_server import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/display/__init__.py [hunk 4]
```diff
@@ -0,0 +1 @@
+from actfw_raspberrypi.display import *
```
Comment (notogawa) [inline actfw/display/__init__.py:1]:
```suggestion
from actfw_raspberrypi.display import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-raspberrypi instead"))
```

File: actfw/edid/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from actfw_raspberrypi.edid import *
```
Comment (notogawa) [inline actfw/edid/__init__.py:1]:
```suggestion
from actfw_raspberrypi.edid import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-raspberrypi instead"))
```

File: actfw/task/__init__.py [hunk 6]
```diff
@@ -1,6 +1 @@
-from .task import Task
-
-from .isolated import Isolated
-from .pipe import Pipe
-from .consumer import Consumer
-from .producer import Producer
+from actfw_core.task import *
```
Comment (notogawa) [inline actfw/task/__init__.py:1]:
```suggestion
from actfw_core.task import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/task/consumer/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+from actfw_core.task.consumer import *
```
Comment (notogawa) [inline actfw/task/consumer/__init__.py:1]:
```suggestion
from actfw_core.task.consumer import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/task/isolated/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+from actfw_core.task.isolated import *
```
Comment (notogawa) [inline actfw/task/isolated/__init__.py:1]:
```suggestion
from actfw_core.task.isolated import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/task/pipe/__init__.py [hunk 9]
```diff
@@ -0,0 +1 @@
+from actfw_core.task.pipe import *
```
Comment (notogawa) [inline actfw/task/pipe/__init__.py:1]:
```suggestion
from actfw_core.task.pipe import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/task/producer/__init__.py [hunk 10]
```diff
@@ -0,0 +1 @@
+from actfw_core.task.producer import *
```
Comment (notogawa) [inline actfw/task/producer/__init__.py:1]:
```suggestion
from actfw_core.task.producer import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/task/task/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+from actfw_core.task.task import *
```
Comment (notogawa) [inline actfw/task/task/__init__.py:1]:
```suggestion
from actfw_core.task.task import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/v4l2/__init__.py [hunk 12]
```diff
@@ -1,2 +1 @@
-from . import video
-from . import types
+from actfw_core.v4l2 import *
```
Comment (notogawa) [inline actfw/v4l2/__init__.py:1]:
```suggestion
from actfw_core.v4l2 import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/v4l2/control/__init__.py [hunk 13]
```diff
@@ -0,0 +1 @@
+from actfw_core.v4l2.control import *
```
Comment (notogawa) [inline actfw/v4l2/control/__init__.py:1]:
```suggestion
from actfw_core.v4l2.control import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/v4l2/types/__init__.py [hunk 14]
```diff
@@ -0,0 +1 @@
+from actfw_core.v4l2.types import *
```
Comment (notogawa) [inline actfw/v4l2/types/__init__.py:1]:
```suggestion
from actfw_core.v4l2.types import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

File: actfw/v4l2/video/__init__.py [hunk 15]
```diff
@@ -0,0 +1 @@
+from actfw_core.v4l2.video import *
```
Comment (notogawa) [inline actfw/v4l2/video/__init__.py:1]:
```suggestion
from actfw_core.v4l2.video import *
from warnings import warn
warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```

**Review by notogawa — CHANGES_REQUESTED** [review]
It is better to display a deprecation warning if a user imports actfw.

File: actfw/task/consumer/__init__.py [hunk 16]
```diff
@@ -0,0 +1,5 @@
...
+from warnings import warn
+warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
+from warnings import warn
+warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```
Comment (notogawa) [inline actfw/task/consumer/__init__.py:5]: dup

File: actfw/task/isolated/__init__.py [hunk 17]
```diff
@@ -0,0 +1,5 @@
...
+from warnings import warn
+warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
+from warnings import warn
+warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```
Comment (notogawa) [inline actfw/task/isolated/__init__.py:5]: dup

File: actfw/task/pipe/__init__.py [hunk 18]
```diff
@@ -0,0 +1,5 @@
...
+from warnings import warn
+warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
+from warnings import warn
+warn(DeprecationWarning("actfw is DEPRECATED. Use actfw-core instead"))
```
Comment (notogawa) [inline actfw/task/pipe/__init__.py:5]: dup

**Review by notogawa — CHANGES_REQUESTED** [review]

File: actfw/task/consumer/__init__.py (hunk 16 above)
Comment (laysakura) [reply actfw/task/consumer/__init__.py]: 😭 ef2ea29

File: actfw/task/isolated/__init__.py (hunk 17 above)
Comment (laysakura) [reply actfw/task/isolated/__init__.py]: ef2ea29

File: actfw/task/pipe/__init__.py (hunk 18 above)
Comment (laysakura) [reply actfw/task/pipe/__init__.py]: ef2ea29

**Review by notogawa — DISMISSED** [review]
LGTM

**Comment by laysakura** [comment]
Thanks.
I will merge this PR after:

- @laysakura Upload release version of actfw-core and actfw-raspberrypi to PyPI after code review.
- @laysakura rerun CI (currently CI fails because actfw-{core,raspberrypi} are pre-release version and cannot install without --pre option.

**Comment by laysakura** [comment]
>  @laysakura Upload actfw v1.4.0a1 to TestPyPI after code review.
>  @laysakura Test forward compatibility with some Actcast applications.
>     using actcast-app-idein-imagenet-classification (private)

finished.

I'm checking CI fail.

**Comment by laysakura** [comment]
> ======================================================================
> ERROR: test_import.test_import_actfw_forward_compatibility:325
> {'from': 'actfw.v4l2.video', 'import': 'macholib'}
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/root/project/actfw/test/test_import.py", line 379, in test_import_actfw_forward_compatibility
>     exec(f'''from {param['from']} import {param['import']}''')
>   File "<string>", line 1, in <module>
> ImportError: cannot import name 'macholib' from 'actfw.v4l2.video' (/root/project/actfw/v4l2/video/__init__.py)

Seemingly because 

1. `ctypes.matcholib` existed in ctypes when I generate test parameters in `test/test_import.py` but
2. `ctypes.matcholib` has been removed in current ctypes.

The point is we don't have to assure import `cytpes` objects from `actfw.**` path.

I will change test parameters to only include Constants we ourselves define in actfw_core package.

**Comment by laysakura** [comment]
@notogawa Please make your review again from this comment: https://github.com/Idein/actcast-app-python/pull/40#issuecomment-769607539

**Review by notogawa — COMMENTED** [review]
Importing tests for symbols in https://github.com/Idein/actfw-core/blob/master/actfw_core/v4l2/control.py and https://github.com/Idein/actfw-core/blob/master/actfw_core/v4l2/types.py are deleted at https://github.com/Idein/actcast-app-python/pull/40/commits/26d312d55f4186f3f48c3f276fe339a438e50073.

Is this what you intended?

**Comment by laysakura** [comment]
ah, actually no 😰 

I should allow importing symbols in `video.py` added by:

```python
from actfw_core.v4l2.types import *
from actfw_core.v4l2.control import *
```

**Comment by laysakura** [comment]
@notogawa added 7710ef2

**Review by notogawa — APPROVED** [review]
LGTM

**Pull request merged**